x-pack/plugins/aiops/public/components/change_point_detection/use_change_point_agg_request.ts

```diff
@@ -0,0 +1,270 @@
+import { useCancellableRequest } from '../../hooks/use_cancellable_request';
+import { useSplitFieldCardinality } from './use_split_field_cardinality';
+
+interface RequestOptions {
+  index: string;
+  fn: string;
+  metricField: string;
+  splitField: string;
+  timeField: string;
+  timeInterval: string;
+  afterKey?: string;
+}
+
+export const COMPOSITE_AGG_SIZE = 500;
+
+function getChangePointDetectionRequestBody(
+  { index, fn, metricField, splitField, timeInterval, timeField, afterKey }: RequestOptions,
+  query: QueryDslQueryContainer
+) {
+  return {
+    params: {
+      index,
+      size: 0,
+      body: {
+        query,
+        aggregations: {
+          groupings: {
+            composite: {
+              size: COMPOSITE_AGG_SIZE,
+              ...(afterKey !== undefined ? { after: { splitFieldTerm: afterKey } } : {}),
+              sources: [
+                {
+                  splitFieldTerm: {
+                    terms: {
+                      field: splitField,
+                    },
+                  },
+                },
+              ],
+            },
+            aggregations: {
+              over_time: {
+                date_histogram: {
+                  field: timeField,
+                  fixed_interval: timeInterval,
+                },
+                aggs: {
+                  function_value: {
+                    [fn]: {
+                      field: metricField,
+                    },
+                  },
+                },
+              },
+              change_point_request: {
+                change_point: {
+                  buckets_path: 'over_time>function_value',
+                },
+              },
+              select: {
+                bucket_selector: {
+                  buckets_path: { p_value: 'change_point_request.p_value' },
+                  script: 'params.p_value < 1',
+                },
+              },
+              sort: {
+                bucket_sort: {
+                  sort: [{ 'change_point_request.p_value': { order: 'asc' } }],
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  };
+}
+
+const CHARTS_PER_PAGE = 6;
+
+export function useChangePointResults(
+  requestParams: ChangePointDetectionRequestParams,
+  query: QueryDslQueryContainer
+) {
+  const {
+    notifications: { toasts },
+  } = useAiopsAppContext();
+
+  const { dataView } = useDataSource();
+
+  const [results, setResults] = useState<ChangePointAnnotation[]>([]);
+  const [activePage, setActivePage] = useState<number>(0);
+  const [progress, setProgress] = useState<number>(0);
+
+  const splitFieldCardinality = useSplitFieldCardinality(requestParams.splitField, query);
+
+  const { runRequest, cancelRequest, isLoading } = useCancellableRequest();
+
+  useEffect(
+    function cancelRequestOnChange() {
+      cancelRequest();
+
+      return () => {
+        cancelRequest();
+      };
+    },
+    [query, cancelRequest, requestParams]
```

Comment:
it appears `cancelRequest` will be called more than once as the `useEffect` on line 219 is also watching `query` and `requestParams` and it calls `reset`.
Is there a risk of a race condition between these useEffects and the other trigger for `fetchResults`?

Reply:
fixed in https://github.com/elastic/kibana/pull/144093/commits/1e0bdabdea409722d3c195f139b14f0c011392e9